### PR TITLE
Implement ACT and ACT with language embedding

### PR DESCRIPTION
**Pull request by LeeHakHo**

### CODE REVIEW

File: examples/baselines/act/act/utils.py [hunk 1]
```diff
@@ -96,14 +96,29 @@ def worker_init_fn(worker_id, base_seed=None):
...
     elif isinstance(file, Dataset):
-        return file[()]
+        try:
+            return file[()]
+        except OSError as e:
+            print(f"[H5 WARN] read failed -> replacing with zeros")
+            print(f"  dataset: {path}")
+            print(f"  shape={file.shape}, dtype={file.dtype}, compression={file.compression}, chunks={file.chunks}")
+            print(f"  err: {e}")
+            return np.zeros(file.shape, dtype=file.dtype)
```
Comment (jstmn) [inline examples/baselines/act/act/utils.py:116]: this part seems sketchy. When would this happen? If there's a file reading error, it likely means the h5 is corrupted. Returning all 0s could result in massive downstream issues

File: examples/baselines/act/act/utils.py [hunk 2]
```diff
@@ -96,14 +96,29 @@ def worker_init_fn(worker_id, base_seed=None):
     # 'rewards': 'rewards',
     'actions': 'actions',
 }
-def load_content_from_h5_file(file):
+
+def load_content_from_h5_file(file, path=""):
```
Comment (jstmn) [inline examples/baselines/act/act/utils.py]: rename to 'prefix' or something else more appropriate plz. 'path' is too similar to filepath

File: examples/baselines/act/inference.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    
+    eval_args = tyro.cli(EvalArgs)
+    
+    train_rgbd.args = eval_args 
+    
+    device = torch.device("cuda" if torch.cuda.is_available() and eval_args.cuda else "cpu")
+
+
+    if not os.path.exists(eval_args.checkpoint_path):
+        raise FileNotFoundError(f"체크포인트를 찾을 수 없습니다: {eval_args.checkpoint_path}")
```
Comment (jstmn) [inline examples/baselines/act/inference.py]: english please

File: examples/baselines/act/train_rgbd.py [hunk 4]
```diff
@@ -34,13 +35,15 @@
...
-from mani_skill.envs.tasks.tabletop import *
+import wandb
+from mani_skill.utils.io_utils import load_json
 
 # Note(@jstmn): 'world__T__ee', 'world__T__root' were added to the observation space of the Panda agent as a 
 # convenience feature. We remove it here because it isn't included in the default ACT method. Additionally, it 
 # causes at torch shape mismatch error, because the configuration space is [batch x ndof], but these values are
 # [batch x 4 x 4]
 OBS_KEYS_TO_REMOVE = {"world__T__ee", "world__T__root"}
+ALLOWED_OBS_EXTRA_KEYS = {"tcp_pose", "left_arm_tcp", "right_arm_tcp", "goal_pos", "is_grasped"}
```
Comment (jstmn) [inline examples/baselines/act/train_rgbd.py]: please remove

File: examples/baselines/act/train_rgbd.py [hunk 5]
```diff
@@ -57,11 +60,11 @@ class Args:
     """if toggled, cuda will be enabled by default"""
     track: bool = False
     """if toggled, this experiment will be tracked with Weights and Biases"""
-    wandb_project_name: str = "ManiSkill"
+    wandb_project_name: str = "ManiSkill_fixed"
```
Comment (jstmn) [inline examples/baselines/act/train_rgbd.py]: rm '_fixed'

File: examples/baselines/act/train_rgbd.py [hunk 6]
```diff
@@ -57,11 +60,11 @@ class Args:
...
     """the wandb's project name"""
     wandb_entity: Optional[str] = None
     """the entity (team) of wandb's project"""
-    capture_video: bool = True
+    capture_video: bool = False #Hayden
```
Comment (jstmn) [inline examples/baselines/act/train_rgbd.py]: rm comment

File: examples/baselines/act/train_rgbd.py [hunk 7]
```diff
@@ -163,25 +178,58 @@ def observation(self, observation: Dict):
...
+        
         if self.include_depth:
             depth = torch.stack(images_depth, dim=1) # (1, num_cams, C, 224, 224), float16
 
         # flatten the rest of the data which should just be state data
         observation = common.flatten_state_dict(observation, use_torch=True)
+
+        #Hayden
+        if args.real:
+            observation = observation[..., :18]
```
Comment (jstmn) [inline examples/baselines/act/train_rgbd.py]: what's happening here? can you leave a comment

File: examples/baselines/act/train_rgbd.py [hunk 8]
```diff
@@ -163,25 +178,58 @@ def observation(self, observation: Dict):
...
+            if cur < self.target_num_cams:
+                pad = torch.zeros(
+                    (rgb.shape[0], self.target_num_cams - cur, rgb.shape[2], rgb.shape[3], rgb.shape[4]),
+                    dtype=rgb.dtype,
+                    device=rgb.device
+                )
+                rgb = torch.cat([rgb, pad], dim=1)
+
+            elif cur > self.target_num_cams:
+                rgb = rgb[:, :self.target_num_cams]
```
Comment (jstmn) [inline examples/baselines/act/train_rgbd.py]: this seems sketchy too. what's going on here?

File: examples/baselines/act/train_rgbd.py [hunk 9]
```diff
@@ -196,7 +244,7 @@ def observation(self, observation: Dict):
 
 
 class SmallDemoDataset_ACTPolicy(Dataset): # Load everything into memory
-    def __init__(self, data_path, num_queries, num_traj, include_depth=True):
+    def __init__(self, data_path, num_queries, num_traj, include_depth=True, is_multi_task=True, target_state_dim=None):
```
Comment (jstmn) [inline examples/baselines/act/train_rgbd.py]: > is_multi_task=True

remove default here

File: examples/baselines/act/train_rgbd.py (hunk 8 above)
Comment (jstmn) [reply examples/baselines/act/train_rgbd.py]: why are we padding the rgb?

File: examples/baselines/act/train_rgbd.py [hunk 10]
```diff
@@ -138,12 +145,14 @@ class FlattenRGBDObservationWrapper(gym.ObservationWrapper):
     Note that the returned observations will have a "rgbd" or "rgb" or "depth" key depending on the rgb/depth bool flags.
     """
 
-    def __init__(self, env, rgb=True, depth=True, state=True) -> None:
+    def __init__(self, env, rgb=True, depth=True, state=True, is_multi_task=True, target_num_cams=3) -> None:
```
Comment (jstmn) [inline examples/baselines/act/train_rgbd.py]: > target_num_cams=3

why is the default 3?

File: examples/baselines/act/train_rgbd.py [hunk 11]
```diff
@@ -221,9 +274,42 @@ def __init__(self, data_path, num_queries, num_traj, include_depth=True):
...
+            self.json_data = load_json(data_path.replace(".h5", ".json"))
+            self.episode_env_ids = []
+            for ep in self.json_data["episodes"]:
+                eid = ep.get("env_id", args.env_id)
+                self.episode_env_ids.append(eid)
+
+            from collections import Counter
+            dims = [o["state"].shape[1] for o in trajectories["observations"]]
+            
+            if self.target_state_dim is None:
```
Comment (jstmn) [inline examples/baselines/act/train_rgbd.py:278]: why would there be different state_dims across the demos? It seems like it would be better if we simply throw an error if the state_dim is different for any two demos

File: examples/baselines/act/train_rgbd.py [hunk 12]
```diff
@@ -221,9 +274,42 @@ def __init__(self, data_path, num_queries, num_traj, include_depth=True):
...
 
+
+        if self.is_multi_task:
+            self.json_data = load_json(data_path.replace(".h5", ".json"))
+            self.episode_env_ids = []
+            for ep in self.json_data["episodes"]:
+                eid = ep.get("env_id", args.env_id)
+                self.episode_env_ids.append(eid)
+
+            from collections import Counter
```
Comment (jstmn) [inline examples/baselines/act/train_rgbd.py]: import to top of file

File: examples/baselines/act/train_rgbd.py [hunk 13]
```diff
@@ -221,9 +274,42 @@ def __init__(self, data_path, num_queries, num_traj, include_depth=True):
...
+            print("[STATE TARGET DIM]", self.target_state_dim )
+
+            for o in trajectories["observations"]:
+                s = o["state"]
+                d = s.shape[1] 
+                if d < self.target_state_dim:
+                    pad = torch.zeros((s.shape[0], self.target_state_dim  - d), dtype=s.dtype)
+                    o["state"] = torch.cat([s, pad], dim=1)
+                elif d > self.target_state_dim:
+                    o["state"] = s[:, :self.target_state_dim ]
```
Comment (jstmn) [inline examples/baselines/act/train_rgbd.py:290]: padding the state also seems sketchy. I feel like we should enforce a single dim for all demos

File: examples/baselines/act/train_rgbd.py [hunk 14]
```diff
@@ -221,9 +274,42 @@ def __init__(self, data_path, num_queries, num_traj, include_depth=True):
...
+        #     #--------------------------
+
         # Pre-process the actions
         for i in tqdm.tqdm(range(len(trajectories['actions'])), desc='Pre-processing actions'):
-            trajectories['actions'][i] = torch.Tensor(trajectories['actions'][i])
+            current_act = torch.Tensor(trajectories['actions'][i])
+
+            #Hayen
+            if args.real:
+                current_act = current_act[..., :8]
```
Comment (jstmn) [inline examples/baselines/act/train_rgbd.py]: ??? this is sketchy. the dimension shouldn't be changed like this. better to instead throw an error if it doesn't match the expected

File: examples/baselines/act/train_rgbd.py [hunk 15]
```diff
@@ -249,6 +335,14 @@ def __init__(self, data_path, num_queries, num_traj, include_depth=True):
         self.delta_control = 'delta' in args.control_mode
         self.norm_stats = self.get_norm_stats() if not self.delta_control else None
 
+
+        #Hayden - multi-task
+        if self.is_multi_task:
+            #self.json_data = load_json(data_path.replace(".h5", ".json"))
+            #self.episode_env_ids = [ep.get("env_id", "Unknown-v1") for ep in self.json_data["episodes"]]
+            from collections import Counter
```
Comment (jstmn) [inline examples/baselines/act/train_rgbd.py]: duplicate import

File: examples/baselines/act/train_rgbd.py [hunk 16]
```diff
@@ -268,6 +362,12 @@ def __getitem__(self, index):
             elif not self.delta_control:
                 target = act_seq[-1]
                 act_seq = torch.cat([act_seq, target.repeat(self.num_queries-action_len, 1)], dim=0)
+            else:
+                #multi_task
+                pad_size = self.num_queries - action_len
+                last_action = act_seq[-1:] # (1, act_dim)
+                padding = last_action.repeat(pad_size, 1)
+                act_seq = torch.cat([act_seq, padding], dim=0)
```
Comment (jstmn) [inline examples/baselines/act/train_rgbd.py]: sketchy also. why are we now repeating the last action?

File: examples/baselines/act/train_rgbd.py [hunk 17]
```diff
@@ -302,28 +402,41 @@ def process_obs(self, obs_dict):
...
+        if self.include_depth:
+            depth_tensor = torch.Tensor(cam_data["depth"].astype(np.float32) / 1024).to(torch.float16)
+            resized_depth = self.transforms(depth_tensor.permute(0, 3, 1, 2))
+            depth = torch.stack([resized_depth], dim=1)
+
+        # obs_dict['extra']
+        if isinstance(obs_dict, dict) and 'extra' in obs_dict:
+            new_extra = {}
+            for k, v in obs_dict['extra'].items():
+                if k in ALLOWED_OBS_EXTRA_KEYS:
```
Comment (jstmn) [inline examples/baselines/act/train_rgbd.py]: rm

File: examples/baselines/act/train_rgbd.py [hunk 18]
```diff
@@ -302,28 +402,41 @@ def process_obs(self, obs_dict):
...
         obs_dict = common.flatten_state_dict(obs_dict, use_torch=True)
+        
+        if self.is_multi_task and (self.target_state_dim is not None):
+            s = obs_dict
+            d = s.shape[1]
+            if d < self.target_state_dim:
+                pad = torch.zeros((s.shape[0], self.target_state_dim - d), dtype=s.dtype)
+                obs_dict = torch.cat([s, pad], dim=1)
+            elif d > self.target_state_dim:
+                obs_dict = s[:, :self.target_state_dim]
```
Comment (jstmn) [inline examples/baselines/act/train_rgbd.py:412]: remember to remove if `target_state_dim` is removed

File: examples/baselines/act/train_rgbd.py [hunk 19]
```diff
@@ -430,6 +557,26 @@ def get_action(self, obs):
...
+                )
+                obs['state'] = torch.cat([obs['state'], padding], dim=-1)
+
+            current_num_cams = obs['rgb'].shape[1]
+            if current_num_cams < self.num_cams:
+                # [B, 1, 3, 224, 224] -> [B, 2, 3, 224, 224]
+                pad_shape = (obs['rgb'].shape[0], self.num_cams - current_num_cams, 3, 224, 224)
+                padding_rgb = torch.zeros(pad_shape, device=obs['rgb'].device, dtype=obs['rgb'].dtype)
+                obs['rgb'] = torch.cat([obs['rgb'], padding_rgb], dim=1)
+
```
Comment (jstmn) [inline examples/baselines/act/train_rgbd.py:539]: why is this happening? seems suspicious we're making this kind of change only for multi-task

File: examples/baselines/act/train_rgbd.py [hunk 20]
```diff
@@ -465,6 +612,15 @@ def save_ckpt(run_name, tag):
...
     assert args.sim_backend in ("physx_cpu", "physx_cuda")
 
+    #multi-task
+    demo_json = None
+    is_multi_task = False
+    try:
+        demo_json = load_json(args.demo_path.replace(".h5", ".json"))
+        is_multi_task = bool(demo_json.get("multi_env", False))
+    except Exception as e:
+        print("[WARN] failed to read demo json for multi_env:", e)
```
Comment (jstmn) [inline examples/baselines/act/train_rgbd.py]: should we throw an error here?

File: examples/baselines/act/train_rgbd.py [hunk 21]
```diff
@@ -540,7 +710,8 @@ def save_ckpt(run_name, tag):
     )
 
     # agent setup
-    agent = Agent(envs, args).to(device)
+    agent = Agent(envs, args, is_multi_task).to(device)
+    #agent = torch.compile(agent) #Hayden
```
Comment (jstmn) [inline examples/baselines/act/train_rgbd.py]: rm if not working

File: examples/baselines/act_clip/act/detr/clip/bpe_simple_vocab_16e6.txt.gz
Comment (jstmn) [inline examples/baselines/act_clip/act/detr/clip/bpe_simple_vocab_16e6.txt.gz:1]: rm

File: examples/baselines/act_clip/baselines.sh
Comment (jstmn) [inline examples/baselines/act_clip/baselines.sh:1]: rm this file

**Comment by jstmn** [comment]
Overall looks good, left some comments

**Review by jstmn — CHANGES_REQUESTED** [review]

File: examples/baselines/act/act/utils.py (hunk 1 above)
Comment (LeeHakHo) [reply examples/baselines/act/act/utils.py:116]: This happens in the dataset at traj_600/obs/sensor_data/hand_camera/rgb. We’ll probably need to collect a new dataset.

File: examples/baselines/act/train_rgbd.py (hunk 7 above)
Comment (LeeHakHo) [reply examples/baselines/act/train_rgbd.py]: In the simulation environment, we included various state information beyond the robot’s 18 states—such as object_state, velocities, and more—but it didn’t match the real Franka robot’s state dimensions, so we discarded it.

File: examples/baselines/act/act/utils.py (hunk 1 above)
Comment (jstmn) [reply examples/baselines/act/act/utils.py:116]: OK. let's remove the np.zeros(file.shape, dtype=file.dtype) part. It should just fail. It's better to fix the actual issue with the h5.

File: examples/baselines/act/train_rgbd.py (hunk 7 above)
Comment (jstmn) [reply examples/baselines/act/train_rgbd.py]: What's the dimension of the observation in real? What are we throwing out? Can you add an assert that the observation matches the expected size.

File: examples/baselines/act/train_rgbd.py (hunk 16 above)
Comment (jstmn) [reply examples/baselines/act/train_rgbd.py]: wait i'm confused now. What exactly is `num_queries`? 

But besides that, why are we changing the default code? Hasn't this been working already?
